### PR TITLE
Increase penalty to ensure constraints are satisfied

### PR DESCRIPTION
**Pull request by mcfarljm**
At each feature selection size, the problem formulation includes a constraint to ensure that the specified number of features are selected.  This is included via a penalty function, but the penalty was not large enough to ensure that the ground state is a feasible solution.  This can easily be verified using the exact solver.  The following table shows each feature size and the corresponding number selected for the ground state, indicating that the constraint is not satisfied for 5 out of the 8 cases:

| n features | n selected |
| ---------- | ---------- |
|          1 |          1 |
|          2 |          2 |
|          3 |          4 |
|          4 |          5 |
|          5 |          6 |
|          6 |          7 |
|          7 |          8 |
|          8 |          8 |

This change increases the value of the penalty to ensure that the constraints are satisfied for the ground state.

In order to make the specification of the penalty value more general, a function has been added to compute the maximum change in energy when flipping a single variable.

### CODE REVIEW

**Comment by mcfarljm** [comment]
@hsadeghidw, requesting review (I don't think I'm able to formally add reviewers)

**Comment by JoelPasvolsky** [comment]
@mcfarljm, it's clear to me that your penalty strength, which is about three times the current value, makes it unlikely the $n \choose k$ penalty will be violated but unclear to me that the objective is still likely to be solved correctly because biases for the objective are on the order of abs(0.1) and abs(2) for linear and quadratic, respectively, so setting a penalty to around 18 means quadratic biases of about abs(1000) for k=8, right? Won't this risk having the weightings between MIs become lost in the imprecision of the sampling?

Can you say more about why you chose the maximum change in energy when flipping a single variable as a good value for the penalty on the $n \choose k$ constraint? I.e., what's the upper boundary that complements the lower bound?

Thanks!

**Comment by mcfarljm** [comment]
This strategy for choosing the penalty is also discussed in Lucas's 2014 paper and is how he determined penalty coefficient ranges for many of the formulations.  If the incremental penalty that you incur when flipping a single variable is larger than the greatest possible benefit (decrease in objective function value), then there is no way for that change to be favored by the augmented objective function.  In other words, it ensures that no change from a feasible to infeasible state can result in an improvement to the augmented objective.  Because the penalty accumulates as further violations are made, then there should be no way for an infeasible solution to have a better augmented objective value than any feasible solution.

Previously, the problem was formulated such that the solution that minimized the augmented objective function (original problem objective plus penalty) was not actually a feasible solution (for some values of k).  IMO, this is an incorrect definition of the problem, because what you are asking the algorithm to try and find (the minimum of the augmented objective) is not actually the solution to the problem that you want to solve (choose k out of n features that maximize the given MI formulation)

The largest absolute value of the linear or quadratic biases that I see is 266.  I would not be able to speak to what magnitudes will start to create problems for the QPU.  However, from the testing that I've done, `DWaveCliqueSampler` finds a feasible solution with this version of the code, whereas it did not with the original version.

**Review by arcondello — APPROVED** [review]

File: titanic.py [hunk 1]
```diff
@@ -73,6 +73,17 @@ def conditional_mutual_information(p, j, *conditional_indices):
     return (conditional_shannon_entropy(np.sum(p, axis=j), *conditional_indices)
             - conditional_shannon_entropy(p, j, *conditional_indices))
 
+def maximum_energy_delta(bqm):
```
Comment (arcondello) [inline titanic.py:76]: This function might make sense to include as a tool in Ocean

Comment (randomir) [reply titanic.py:76]: If this finds its way to Ocean, it should be generalized to handle ising bqms (`delta *= 2` after summations)

Comment (arcondello) [reply titanic.py:76]: Also, can be made much more performant by cythonizing it.

**Pull request merged**